File: docs/test-cases/cryptographic-failures.go
// Test case: cryptographic-failures (A02:2025)
package vault

import (
	"crypto/aes"
	"crypto/des"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"math/rand"
)

// BUG: hardcoded encryption key in source control
var encryptionKey = []byte("hardcoded-key-16")

func HashPassword(password string) string {
	// BUG: MD5 is broken for password hashing — use bcrypt/argon2
	sum := md5.Sum([]byte(password))
	return hex.EncodeToString(sum[:])
}

func NewSessionToken() string {
	// BUG: math/rand is not cryptographically secure — use crypto/rand
	b := make([]byte, 32)
	for i := range b {
		b[i] = byte(rand.Intn(256))
	}
	return hex.EncodeToString(b)
}

func EncryptLegacy(plaintext []byte) ([]byte, error) {
	// BUG: DES is obsolete (56-bit key, brute-forceable)
	block, err := des.NewCipher([]byte("8bytekey"))
	if err != nil {
		return nil, err
	}
	out := make([]byte, len(plaintext))
	block.Encrypt(out, plaintext)
	return out, nil
}

func EncryptECB(plaintext []byte) ([]byte, error) {
	block, err := aes.NewCipher(encryptionKey)
	if err != nil {
		return nil, err
	}
	// BUG: ECB mode leaks plaintext structure — use AES-GCM
	out := make([]byte, len(plaintext))
	for i := 0; i < len(plaintext); i += block.BlockSize() {
		block.Encrypt(out[i:i+block.BlockSize()], plaintext[i:i+block.BlockSize()])
	}
	fmt.Println("encrypted")
	return out, nil
}

File: docs/test-cases/cryptographic-failures.rs
// Test case: cryptographic-failures (A02:2025)
use md5::{Digest, Md5};
use rand::Rng;

// BUG: hardcoded encryption key bytes committed to source
const ENCRYPTION_KEY: [u8; 16] = [
    0x68, 0x61, 0x72, 0x64, 0x63, 0x6f, 0x64, 0x65,
    0x64, 0x6b, 0x65, 0x79, 0x31, 0x32, 0x33, 0x34,
];

pub fn hash_password(password: &str) -> String {
    // BUG: MD5 is broken and unsalted — use argon2 or bcrypt
    let mut hasher = Md5::new();
    hasher.update(password.as_bytes());
    format!("{:x}", hasher.finalize())
}

pub fn generate_api_secret() -> String {
    // BUG: rand::random() uses ThreadRng — for secrets use OsRng / getrandom
    let bytes: [u8; 32] = rand::random();
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

pub fn weak_session_id() -> u64 {
    // BUG: thread_rng is not guaranteed cryptographic — use OsRng for tokens
    let mut rng = rand::thread_rng();
    rng.gen::<u64>()
}

pub fn encrypt_ecb(plaintext: &[u8]) -> Vec<u8> {
    use aes::cipher::{generic_array::GenericArray, BlockEncrypt, KeyInit};
    use aes::Aes128;
    let cipher = Aes128::new(GenericArray::from_slice(&ENCRYPTION_KEY));
    let mut block = GenericArray::clone_from_slice(&plaintext[..16]);
    // BUG: raw single-block AES with no mode == ECB; leaks patterns, no integrity
    cipher.encrypt_block(&mut block);
    block.to_vec()
}
